core/src/data/value/mod.rs

```diff
@@ -555,6 +555,157 @@ impl Value {
         }
     }
 
+    pub fn validate_bitwise_shift_rhs(rhs: &Value) -> Result<u32> {
+        use Value::*;
+
+        match rhs {
+            U32(rhs_u32) => Ok(rhs_u32.to_owned()),
+            I64(rhs_i64) => match rhs_i64.to_owned().try_into() {
+                Ok(u) => Ok(u),
+                Err(_) => Err(ValueError::ImpossibleCast.into()),
+            },
+            _ => Err(ValueError::ImpossibleCast.into()),
+        }
+    }
+
+    pub fn bitwise_shift_left(&self, rhs: &Value) -> Result<Value> {
+        use Value::*;
+
+        let try_rhs_u32 = Self::validate_bitwise_shift_rhs(rhs);
+        match try_rhs_u32 {
```

Comment:
we can simplify the codes like these below,
`validate_bitwise_shift_rhs` is now only being used in a single place so no more as a separate function is required.

```rust
let rhs = u32::try_from(rhs)?;
match self {
    Value::I8(lhs) => lhs.checked_shl(rhs).map(Value::I8),
    Value::I16(lhs) => lhs.checked_shl(rhs).map(Value::I16),
    Value::I32(lhs) => lhs.checked_shl(rhs).map(Value::I32),
    Value::I64(lhs) => lhs.checked_shl(rhs).map(Value::I64),
    Value::I128(lhs) => lhs.checked_shl(rhs).map(Value::I128),
    Value::U8(lhs) => lhs.checked_shl(rhs).map(Value::U8),
    Value::U16(lhs) => lhs.checked_shl(rhs).map(Value::U16),
    Value::U32(lhs) => lhs.checked_shl(rhs).map(Value::U32),
    Value::U64(lhs) => lhs.checked_shl(rhs).map(Value::U64),
    Value::U128(lhs) => lhs.checked_shl(rhs).map(Value::U128),
    Value::Null => Some(Value::Null),
    _ => {
        return Err(ValueError::NonNumericMathOperation {
            lhs: self.clone(),
            rhs: Value::U32(rhs),
            operator: NumericBinaryOperator::BitwiseShiftLeft,
        }
        .into());
    }
}
.ok_or_else(|| {
    ValueError::BinaryOperationOverflow {
        lhs: self.clone(),
        rhs: Value::U32(rhs),
        operator: NumericBinaryOperator::BitwiseShiftLeft,
    }
    .into()
})
```